llvm/lib/Transforms/IPO/DeadArgumentElimination.cpp

```diff
@@ -1260,13 +1261,74 @@ void DeadArgumentEliminationPass::UpdateNVPTXMetadata(Module &M, Function *F,
     const auto &FuncOperand = MetadataNode->getOperand(0);
     if (!FuncOperand)
       continue;
-    auto FuncConstant = dyn_cast<ConstantAsMetadata>(FuncOperand);
+    auto *FuncConstant = dyn_cast<ConstantAsMetadata>(FuncOperand);
     if (!FuncConstant)
       continue;
     auto *Func = dyn_cast<Function>(FuncConstant->getValue());
     if (Func != F)
       continue;
     // Update the metadata with the new function
     MetadataNode->replaceOperandWith(0, llvm::ConstantAsMetadata::get(NF));
+
+    // Carefully update grid_constant annotations, since those are denoted
+    // parameter indices, which may have changed
+    if (auto *Type = dyn_cast<MDString>(MetadataNode->getOperand(1));
```

Comment:
Didn't realise that there could be multiple annotations per node a la `!{ptr @grid_const_int, !"kernel", i32 1, !"grid_constant", !1}`. This will need to be updated to account for that.